frb_rust/src/dart_api_dl/include/dart_api_dl.c

```diff
@@ -0,0 +1,59 @@
+/*
```

Comment:
yes, we must, because we compile it

Reply:
Hmm, could you please add a link (e.g. point to official Dart doc) as a comment to somewhere in the codebase about how to use dart_api_dl.c? At a first glance, I feel it is buggy since if two third-party libs compile the same `.c` there can be conflicts etc.

Reply:
from dart_api_dl.h
 ```
 * All symbols are postfixed with _DL to indicate that they are dynamically
 * linked and to prevent conflicts with the original symbol.
 *
 * Link `dart_api_dl.c` file into your library and invoke
 * `Dart_InitializeApiDL` with `NativeApi.initializeApiDLData`.
 ```
 I will search further

Reply:
Theoretically, the error can only be when calling Dart_InitializeApiDL multiple times, but this method returns 0 if ok, which we check. Also I couldn't find any restrictions in the comments

Reply:
Hmm I have not written C code for a while, but IIRC, "link" does not mean "compile"?

* "compile dart_api_dl.c" := compile the source code
* "link dart_api_dl.c" := do not compile that .c file; instead, only "link" it.

For example, we commonly say "link the `libc` library with our Rust or C code". It does not mean "download and compile the C code of the whole libc", but instead means "just fetch something like libc.a and link that precompiled thing"

Reply:
> At a first glance, I feel it is buggy since if two third-party libs compile the same .c there can be conflicts etc.
> the error can only be when calling Dart_InitializeApiDL multiple times

Well I do not mean this. For simplicity, say there are two bare-metal C libraries and they both contain dart_api_dl.c to compile. Then, they will both expose the symbol Dart_InitializeApiDL in their .so files. Therefore, at runtime, the same symbol can come from two places. This seems to be undefined behavior or bad behavior IIRC.

Reply:
Hmm I see https://github.com/dart-lang/sdk/issues/47356#issuecomment-1056770450, so seems the original approach is correct. But I am still quite confused about it:

1. Why we include "internal" things
2. This worry https://github.com/fzyzcjy/flutter_rust_bridge/pull/853#discussion_r1034665554

As said above, I have not written C code for a while, so please correct me if I am wrong!

Reply:
we can just change dart_api_dl.
instead of `Dart_InitializeApiDL`, `Dart_HandleFromPersistent` symbols, we will compile as `Dart_InitializeApiDL_FRB` (basically just change the macro). We will also remove all unnecessary.

also calling Dart_InitializeApiDL_FRB and third party Dart_InitializeApiDL won't conflict in any way

Reply:
Ok I did a bit more research and looks like this (weird) way looks like what people really uses. Then maybe we just keep with your original simplest approach.

> we will compile as Dart_InitializeApiDL_FRB

The worry 2 is not completely removed IMHO, since some people mention they have two crates using frb in one app. But given we have not supported that case it looks acceptable.

Reply:
By the way, I still suggest creating an issue in Dart repo, such that if the official team replies, we can know more about the problem

Reply:
Btw, here is an example about symbol duplication (though very different from our cases): https://stackoverflow.com/questions/55191553/what-actually-happens-when-two-shared-libraries-define-the-same-symbol

> I recently encountered a crash issue when I linked two shared libraries (both made by myself) together. I eventually found it was because of one source file duplicated between the two files. In that source file a global std::vector was defined (in fact a static member of a class), and it ended up with being freed twice -- one by each library.

Again I am not expert in C given lack of using it recently, so feel free to point out my problems!

Reply:
https://github.com/dart-lang/sdk/issues/47356#issuecomment-1331894796
If I understand correctly, then in our case everything is fine (if i asked what you mean)

Reply:
I additionally checked in cases where different libraries call initApiDl and the case when there is a frb dependency of different versions. It didn't give me any errors.